File: GanAndroidApp/app/src/main/java/com/xattacker/gan/service/AccountService.java
package com.xattacker.gan.service;

import com.xattacker.gan.GanAgent;
import com.xattacker.gan.data.FunctionType;
import com.xattacker.gan.data.ResponsePack;
import com.xattacker.binary.BinaryBuffer;

public abstract class AccountService extends ServiceFoundation
{
	public interface AccountServiceListener
	{
		void onLoginSucceed(String aAccount, String aSessionId);
		void onLogouted(String aAccount);
	}
	
	
	private String _account;
	private AccountServiceListener _listener;
	
	protected AccountService(GanAgent aAgent, AccountServiceListener aListener)
	{
		super(aAgent);
		
		_account = null;
		_listener = aListener;
	}
	
	public boolean login(String aAccount, String aPassword)
	{
		boolean result = false;

		try
		{
			BinaryBuffer buffer = new BinaryBuffer();
			buffer.writeString(aAccount);
			buffer.writeString(aPassword);
			
			ResponsePack response = send(FunctionType.LOGIN, buffer.getData());
			if (response != null)
			{
				result = response.getResult();
				
				if (result && _listener != null)
				{
					String session_id = new String(response.getContent());
					_account = aAccount;
					_listener.onLoginSucceed(aAccount, session_id);
				}
				
				response = null;
			}
			
			buffer = null;
		}
		catch (Exception ex)
		{
			android.util.Log.i("aaa", ex.toString());
		}
		
		return result;
	}
	
	public boolean logout()
	{
		boolean result = false;
		
		if (_account != null)
		{
			try
			{
				BinaryBuffer buffer = new BinaryBuffer();
				buffer.writeString(_account);
	
				ResponsePack response = send(FunctionType.LOGOUT, buffer.getData());
				if (response != null)
				{
					result = response.getResult();
					
					if (result && _listener != null)
					{
						_listener.onLogouted(_account);
						_account = null;
					}
				}
				
				buffer = null;
			}
			catch (Exception ex)
			{
				android.util.Log.i("aaa", ex.toString());
			}
		}
		
		return result;
	}
	
	public boolean isAccountExisted(String aAccount)
	{
		boolean existed = false;
		
		
		return existed;
	}
	
	public boolean registerAccount
	(
	String aAccout, 
	String aPassword, 
	byte[] aExtra
	)
	{
		boolean result = false;
		
		try
		{
			BinaryBuffer buffer = new BinaryBuffer();
			buffer.writeString(aAccout);
			buffer.writeString(aPassword);
			
			if (aExtra != null && aExtra.length > 0)
			{
				int size = aExtra.length;
				buffer.writeInteger(size);
				buffer.writeBinary(aExtra, 0, size);
			}
			else
			{
				buffer.writeInteger(0);
			}

			ResponsePack response = send(FunctionType.REGISTER_ACCOUNT, buffer.getData());
			if (response != null)
			{
				result = response.getResult();
			}
			
			buffer = null;
		}
		catch (Exception ex)
		{
			android.util.Log.i("aaa", ex.toString());
		}
		
		return result;
	}
}

File: GanAndroidApp/GanClient/src/main/java/com/xattacker/gan/service/AccountService.kt
package com.xattacker.gan.service

import android.util.Log
import com.xattacker.binary.BinaryBuffer
import com.xattacker.gan.GanAgent
import com.xattacker.gan.data.FunctionType
import java.net.Socket


data class SessionInfo(val account: String, val sessionId: String)


internal interface AccountServiceListener
{
    fun onLoginSucceed(session: SessionInfo, connection: Socket)
    fun onLoggedOut(account: String)
}


class AccountService internal constructor(agent: GanAgent, private val _listener: AccountServiceListener) : ServiceFoundation(agent)
{
    fun login(account: String, password: String): Boolean
    {
        var result = false
        try
        {
            val buffer = BinaryBuffer()
            buffer.writeString(account)
            buffer.writeString(password)

            val response = send(FunctionType.LOGIN, buffer.data, false)
            if (response != null)
            {
                result = response.result
                if (result && response.response != null)
                {
                    val session_id = String(response.response!!)
                    val session = SessionInfo(account, session_id)
                    _listener.onLoginSucceed(session, response.connection!!)
                }
            }
        }
        catch (ex: Exception)
        {
            Log.i("aaa", ex.toString())
        }

        return result
    }

    fun logout(): Boolean
    {
        var result = false

        try
        {
            this.agent.account?.let {
                account ->
                    val buffer = BinaryBuffer()
                    buffer.writeString(account)

                    val response = send(FunctionType.LOGOUT, buffer.data)
                    if (response != null && response.result)
                    {
                        _listener.onLoggedOut(account)
                        result = true
                    }
            }
        }
        catch (ex: Exception)
        {
            Log.i("aaa", ex.toString())
        }

        return result
    }

    fun isAccountExisted(account: String): Boolean
    {
        return false
    }

    fun registerAccount(account: String, password: String, extra: ByteArray? = null): Boolean
    {
        var result = false
        try
        {
            val buffer= BinaryBuffer()
            buffer.writeString(account)
            buffer.writeString(password)

            if (extra != null && extra.size > 0)
            {
                val size = extra.size
                buffer.writeInteger(size)
                buffer.writeBinary(extra, 0, size)
            }
            else
            {
                buffer.writeInteger(0)
            }

            val response = send(FunctionType.REGISTER_ACCOUNT, buffer.data)
            result = response?.result ?: false
        }
        catch (ex: Exception)
        {
            Log.i("aaa", ex.toString())
        }

        return result
    }
}